### PR TITLE
Introduce sample batch domain model

### PR DESCRIPTION
**Pull request by sven1103**
Adds implementation and support for sample registration and batch registration.

Also includes a new domain dispatcher implementation backed by JobRunr, that allows for persistent
event management.

### CODE REVIEW

File: projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/sample/event/BatchRegistered.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  @Serial
+  private static final long serialVersionUID = 580378782496926484L;
+
+  private final BatchId batchId;
+
+  private final Instant occurredOn;
+
+  private BatchRegistered(Instant occurredOn, BatchId batchId) {
+    this.occurredOn = occurredOn;
+    this.batchId = batchId;
```
Comment (KochTobi) [inline projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/sample/event/BatchRegistered.java:26]: null-checks missing for Instant and BatchId

File: projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/sample/event/SampleRegistered.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  private final Instant occurredOn;
+
+  private final BatchId assignedBatch;
+
+  private final SampleId registeredSample;
+
+  private SampleRegistered(Instant occurredOn, BatchId assignedBatch, SampleId registeredSample) {
+    this.occurredOn = occurredOn;
+    this.assignedBatch = assignedBatch;
+    this.registeredSample = registeredSample;
```
Comment (KochTobi) [inline projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/sample/event/SampleRegistered.java:31]: please add null checks in the constructor

File: projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/sample/Sample.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  private ExperimentId experimentId;
+
+  @Embedded
+  private SampleOrigin sampleOrigin;
+
+  @Embedded
+  private BiologicalReplicateId biologicalReplicateId;
+
+  private Sample(
+      SampleId id, BatchId assignedBatch, String label, ExperimentId experimentId,
```
Comment (KochTobi) [inline projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/sample/Sample.java:43]: That's a lot of parameters. Maybe https://learning.oreilly.com/library/view/refactoring-improving-the/0201485672/ch03.html#:-:text=Long%20Parameter%20List we can think about introducing parameter objects here.

File: projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/sample/Sample.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    private SampleOrigin sampleOrigin;
+
+    @Embedded
+    private BiologicalReplicateId biologicalReplicateId;
+
+    private Sample(
+            SampleId id, BatchId assignedBatch, String label, ExperimentId experimentId, Long experimentalGroupId, SampleOrigin sampleOrigin
+            , BiologicalReplicateId replicateReference
+    ) {
+        this.id = id;
```
Comment (KochTobi) [inline projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/sample/Sample.java]: Please check for null for each argument :)

File: projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/sample/Sample.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+  private Long experimentalGroupId;
+
+  @Embedded
+  private ExperimentId experimentId;
+
+  @Embedded
+  private SampleOrigin sampleOrigin;
+
+  @Embedded
```
Comment (KochTobi) [inline projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/sample/Sample.java:39]: Is there a logical ordering of the fields? If not I would welcome a logical ordering or alphabetic.

File: projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/sample/SampleOrigin.java [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+import life.qbic.projectmanagement.domain.project.experiment.vocabulary.Specimen;
+
+/**
+ * <b><class short description - 1 Line!></b>
+ *
+ * <p><More detailed description - When to use, what it solves, etc.></p>
+ *
+ * @since <version tag>
+ */
+public record SampleOrigin(Species species, Specimen specimen, Analyte analyte) {
```
Comment (KochTobi) [inline projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/sample/SampleOrigin.java:14]: :+1: very good.

You could check in the canonical constructor that none of these are null.

File: projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/service/SampleDomainService.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+   * @param biologicalReplicateId the associated biological replicate
+   * @param sampleOrigin          sample origin information
+   * @return A {@link Result} object with the created {@link Sample} or an error with
+   * {@link ResponseCode}
+   * @since 1.0.0
+   */
+  public Result<Sample, ResponseCode> registerSample(String label, BatchId assignedBatch,
+      ExperimentId experimentId,
+      long experimentalGroupId, BiologicalReplicateId biologicalReplicateId,
+      SampleOrigin sampleOrigin) {
```
Comment (KochTobi) [inline projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/service/SampleDomainService.java:51]: Please put every parameter in one separate line or put as many as possible in one line. (If you use auto-formatting, sometimes linebreaks are not removed. Try to remove all line breaks and format the line again.)

File: projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/service/SampleDomainService.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    var sample = Sample.create(label, assignedBatch, experimentId, experimentalGroupId,
+        biologicalReplicateId,
+        sampleOrigin);
+    Result<Sample, ResponseCode> result = this.sampleRepository.add(sample);
+
+    // For successful registration transactions we dispatch the event
+    result.onValue(createdSample -> DomainEventDispatcher.instance()
+        .dispatch(
+            SampleRegistered.create(createdSample.assignedBatch(), createdSample.sampleId())));
```
Comment (KochTobi) [inline projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/service/SampleDomainService.java:60]: I like it :+1:
```suggestion
    result.map(createdSample -> SampleRegistered.create(createdSample.assignedBatch(), createdSample.sampleId()))
    .onValue(event -> DomainEventDispatcher.instance()
        .dispatch(event));
```

File: projectmanagement/src/test/groovy/life/qbic/projectmanagement/domain/project/service/SampleDomainServiceSpec.groovy [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+            }
+        }
+        DomainEventDispatcher.instance().subscribe(sampleRegistered)
+
+        when:
+        Result<Sample, SampleDomainService.ResponseCode> result = sampleDomainService.registerSample("test sample", BatchId.create(), ExperimentId.create(), 1L, BiologicalReplicateId.create(), new SampleOrigin(new Species("test"), new Specimen("test"), new Analyte("test")))
+
+        then:
+        sampleRegistered.batchIdOfEvent.equals(result.getValue().assignedBatch())
+        sampleRegistered.sampleIdOfEvent.equals(result.getValue().sampleId())
```
Comment (KochTobi) [inline projectmanagement/src/test/groovy/life/qbic/projectmanagement/domain/project/service/SampleDomainServiceSpec.groovy:63]: I think you can replace the subscriber by a mock and check the argument provided to the mocked method. https://spockframework.org/spock/docs/2.3/interaction_based_testing.html#_argument_constraints

**Review by KochTobi — CHANGES_REQUESTED** [review]
Overall looks good. I like that there are tests :+1:
Some small suggestions mainly about null safety

File: projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/sample/Sample.java (hunk 3 above)
Comment (sven1103) [reply projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/sample/Sample.java:43]: Good Idea!

File: projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/sample/SampleRegistrationRequest.java [hunk 10]
```diff
@@ -21,4 +22,14 @@ public record SampleRegistrationRequest(String label, BatchId assignedBatch,
                                         BiologicalReplicateId replicateReference,
                                         SampleOrigin sampleOrigin) {
 
+  public SampleRegistrationRequest(String label, BatchId assignedBatch, ExperimentId experimentId,
```
Comment (KochTobi) [inline projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/sample/SampleRegistrationRequest.java:25]: As those are a lot of parameters, you can use records compact constructor and simply remove the (parameters)
https://blogs.oracle.com/javamagazine/post/java-record-compact-canonical-constructor

File: projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/sample/Sample.java [hunk 11]
```diff
@@ -26,12 +26,14 @@ public class Sample {
   @Embedded
   private BatchId assignedBatch;
   @Embedded
+  @AttributeOverride(name = "id", column = @Column(name = "bio_replicate_id"))
   private BiologicalReplicateId biologicalReplicateId;
   @Embedded
   @AttributeOverride(name = "id", column = @Column(name = "experiment_id"))
```
Comment (KochTobi) [inline projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/sample/Sample.java:32]:
```suggestion
  @AttributeOverride(name = "uuid", column = @Column(name = "experiment_id"))
```

File: projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/sample/Sample.java [hunk 12]
```diff
@@ -26,12 +26,14 @@ public class Sample {
...
   private ExperimentId experimentId;
   private Long experimentalGroupId;
   @EmbeddedId
+  @AttributeOverride(name = "id", column = @Column(name = "sample_id"))
   private SampleId id;
```
Comment (KochTobi) [inline projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/sample/Sample.java:37]:
```suggestion
  @AttributeOverride(name = "uuid", column = @Column(name = "sample_id"))
```

File: projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/sample/Sample.java [hunk 13]
```diff
@@ -26,12 +26,14 @@ public class Sample {
   @Embedded
   private BatchId assignedBatch;
```
Comment (KochTobi) [inline projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/sample/Sample.java:27]:
``` suggestion
  @AttributeOverride(name = "uuid", column = @Column(name = "batch_id"))
  private BatchId assignedBatch;
```

**Review by KochTobi — CHANGES_REQUESTED** [review]
Sadly you have to override the fields in the embeddable.

File: projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/sample/Sample.java [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+  private BatchId assignedBatch;
+  @Embedded
+  @AttributeOverride(name = "id", column = @Column(name = "bio_replicate_id"))
+  private BiologicalReplicateId biologicalReplicateId;
+  @Embedded
+  @AttributeOverride(name = "id", column = @Column(name = "experiment_id"))
+  private ExperimentId experimentId;
+  private Long experimentalGroupId;
+  @EmbeddedId
+  @AttributeOverride(name = "id", column = @Column(name = "sample_id"))
```
Comment (KochTobi) [inline projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/sample/Sample.java:36]:
```suggestion
  @AttributeOverride(name = "uuid", column = @Column(name = "sample_id"))
```

File: projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/sample/event/BatchRegistered.java (hunk 1 above)
Comment (KochTobi) [reply projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/sample/event/BatchRegistered.java]: batchid can be null

**Review by KochTobi — APPROVED** [review]

**Review by KochTobi — CHANGES_REQUESTED** [review]
I am missing some null checks in Sample.java

File: projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/sample/Batch.java [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+   * Creates an empty batch.
+   * <p>
+   * The batch is NOT marked as a pilot batch and considered as a standard batch.
+   *
+   * @param label a human-readable semantic label of the batch
+   * @return the created batch
+   * @since 1.0.0
+   */
+  public static Batch create(String label) {
+    return create(label, false);
```
Comment (KochTobi) [inline projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/sample/Batch.java:56]: Is the label optional?

**Review by KochTobi — CHANGES_REQUESTED** [review]

**Pull request merged**